0.3.0 API

jballanc opened the pull request:
Updated API for v0.3.0 that allows for efficient single-call decode of byte arrays or files to either a pre-allocated `ByteBuffer` or returned as a byte array.

chris-allan committed:
Initial version of 0.3.0 API

sbesson committed:
Modify the build.gradle to use the OME Maven and staging repositories

sbesson committed:
Use the correct terminology for Maven repositories

jballanc committed:
Implement new methods for v0.3.0 API

jballanc committed:
Move debug-compensating fseek

The file pointer needs to be reset before initializing the stream. Before this
fix, compiling with `DEBUG=1` would lead to a crash on any attempt to transcode
files.

jballanc committed:
Get the correct full file URL for test files

jballanc committed:
Try testing the big file first

With the v0.3.0 API changes, we seem to be hitting the limits of how many
directly allocated `ByteBuffer`s one can work with in a single JVM. This moves
the largest file first in the test sequence so we can deal with allocation
issues sooner.

jballanc committed:
Recycle ByteBuffer instances in Java tests

jballanc committed:
Comment out tests checking non-direct-allocated ByteBuffers

jballanc committed:
Disable all ByteBuffer tests

They eat up memory and make everything sad...

chris-allan commented:
@melissalinkert should have a look at see what she thinks of the new API. Is there a particular reason the ByteBuffer tests were disabled in dcf512a, @jballanc?

jballanc commented:
Yes, `ByteBuffer` tests were failing due to the issues we saw before with creating too many `ByteBuffer`s. I suspect this may have to do with the changes to `getImageMetadata`, i.e. there's twice the number of `ByteBuffer`s in play because we're no longer assigning the `ByteBuffer` to an instance and re-using it, but rather passing it as a parameter.

One option would be to change the API specifically for `ByteBuffer`s such that we can instantiate `DecodeContext` with a `ByteBuffer` and re-use it, but then the API for `ByteBuffer` would be different than for `File` or `byte[]` which seems less than idea.

I asked @melissalinkert if support for `ByteBuffer`s was critical, and she indicated it was not, so I had intended to simply remove those bits and simplify the API. Working on a proposal for than in a branch of this branch.

chris-allan commented on java/src/test/java/ome/jxrlib/TestInMemoryDecode.java:
```diff
@@ -99,27 +131,31 @@ public void testByteArray(
         Assert.assertEquals(md5, md5(destination));
     }
 
+    /*
```
Let's add a TestNG annotation here and comment as disabled for now or adjust the test cases to re-use a ByteBuffer if it's a memory issue we're running into.

chris-allan commented on java/src/test/resources/testng.xml:
```diff
@@ -3,6 +3,19 @@
 
 <suite name="ome.jxrlib">
 
+  <test name="FL_3channel_noZ_compression_lossless">
```
Is there a particular reason this was moved up?

jballanc replied:
It's the largest of the test files, so moving it was moved up in the hope that (combined with the buffer reallocation workaround in e29028c) running it first would avoid the issues with `ByteBuffer`s. (This worked on my machine, but apparently not on Travis.)

chris-allan commented:
I can't see a reason not to just make it work by adjusting the test cases. We know the size of the required buffers as well based on the test case fixture data being passed in and we can pre-allocate, pass in and/or reuse the buffer as required by setting the TestNG scope.

jballanc commented:
We could certainly re-jigger the tests to avoid excessive `ByteBuffer` allocation. I already attempted a janky solution along these lines in e29028c, but apparently to no avail. If keeping the `ByteBuffer` APIs around is a priority, I can look at this again, but my impression was that those APIs were work-arounds until direct decoding of `byte[]` was possible (which it is now), and so they aren't really needed any longer.

Or, rather, it's not necessary to use `ByteBuffer`s with `DecodeContext`. We could certainly keep the `ByteBuffer` API in `ImageDecode` and `CodecFactory`, if desired, for those that want to separate `ImageDecode` creation and image decoding. In that case, though, we should write tests that exercise those APIs directly (and avoid more `ByteBuffer`s than necessary).

chris-allan commented:
They're not really a work around but rather an alternative.

The first thing I would try to allocate the buffers ahead of time in the test case itself and utilise TestNG scoping to ensure that they are not destroyed.

sbesson committed:
Explicitly set xcode version

jballanc committed:
Add documentation for C++ classes

This commit includes the Doxygen configuration, a `make` target to generate
docs, and documentation for most of the classes.

jballanc committed:
Fix name of PCRE lib for Travis

jballanc committed:
Add documentation for remaining C++ classes

This commit includes documentation comments for `ImageEncoder`, `ImageMetadata`,
`Resolution`, and `Stream`.

jballanc committed:
Revert back to passing ImageDecoder pointers

For whatever reason it seems that passing object references causes crashes when
using Java `ByteBuffer`s to pass image data. Reverting this back to pass
pointers.

chris-allan committed:
Revert "Recycle ByteBuffer instances in Java tests"

This reverts commit e29028caf72905e084ea7b2730f270c8a5c91219.

chris-allan committed:
Revert "Comment out tests checking non-direct-allocated ByteBuffers"

This reverts commit fe6b4dd192a6f7b5699931afee6ba9a9c48b11d9.

chris-allan committed:
Revert "Disable all ByteBuffer tests"

This reverts commit dcf512ae863ea22b94f2bf844720439b07cfc12a.

chris-allan committed:
Ensure wrapping with a Java exception handler

chris-allan committed:
Fix offset typo

chris-allan committed:
Fix invalid input test cases

jballanc committed:
Fix buffer length parameter typo

chris-allan committed:
Push crash logs if they exist

chris-allan committed:
Clean up long lines

chris-allan committed:
Fix additional buffer length parameter typo

chris-allan commented:
Two things, @jballanc:

 * Lets get some test cases added for decoding into a larger `ByteBuffer` at a given offset
 * We should probably have `getImageMetadata(ByteBuffer source)` added to the API and tests added for it

chris-allan committed:
Clean up long lines

chris-allan committed:
Revert "Revert back to passing ImageDecoder pointers"

This reverts commit f16c98d0d35b1f72e66704c4ad887f0a3c714300.

jballanc committed:
Add an implementation of getImageMetadata for ByteBuffers

jballanc committed:
Add more checks on offset and length arguments

melissalinkert commented:
Java API looks good in general.  I could see it being useful to have something like ```AbstractDecode.decodeFrame(int frame, byte[] source, int sourceOffset, int sourceLength, byte[] destination, int destinationOffset)```, especially for consistency with the other ```decodeFrame``` signatures that take a ```File``` or ```ByteBuffer```.  It might also be helpful to have place-holder ```getFrameCount``` methods that just returns 1 - we know that there is only one frame in practice, but that won't be obvious to anyone external.  Adding a frame count field to ```ImageMetadata``` would also work, if that's easier.

Otherwise, I think it would be good to do a little more bounds checking in the ```decodeFrame``` methods, to ensure that the frame and offset are non-negative and the length is positive.

jballanc committed:
Unify API for NIO Buffers and byte[]

This is the first step of refactoring the C++ and Java APIs so that they are
more consistent. On the C++ side, input buffers are always `unsigned char` and
instead of multiple overloads, optional offsets are represented as optional
parameters. This does necessitate rearranging parameter order on some methods,
so all methods were re-arranged so that offsets and lengths appear at the end of
the parameter list.

On the Java side, instead of relying on the built-in SWIG mechanisms for passing
NIO Buffers and `byte[]` across the JNI bridge, we have copied the definitions
into our `JXR.i` definitions and changed the `byte[]` implementation to pass
across as `unsigned char *`. This allows us to use the same C++ methods for
both, though due to limitations with SWIG auto-generation, we need to use the
`%rename` directive to prevent definitions from replacing each other. This
results in a new set of more descriptive method names.

jballanc committed:
Rectify documentation with revised method signatures

This change brings the header doc comments back in line with the actual
parameter lists for DecodeContext's methods that were changed in the last commit.

jballanc committed:
Add frameCount to ImageMetadata

jballanc committed:
Update AbstractDecode Java class to reflect refactored C++ API

With this set of changes, we should have the final version of the API that we
want for AbstractDecode, including congruent methods to decode either `byte[]`s
or `ByteBuffer`s with an offset and length.

jballanc committed:
Remove "toBytes" SWIG methods

These methods gave the API just a bit *too much* surface area. For now, if a
user needs a `byte[]` from the decoded image, they can use the methods that
return `byte[]` and if they need to write into an array at an offset they can
use the methods that write to `ByteBuffer`s. (Really, we just don't have
sufficient testing around passing `byte[]`s across the SWIG/JNI bridge as out
parameters.)

jballanc committed:
Remove un-used classes from SWIG

With the finished `AbstractDecode`/`Decode` API, we no longer need mappings to
the lower-level C++ APIs. Not to mention that, due to issues with pointer
lifetimes, we cannot guarantee their use from Java is safe.

jballanc committed:
Update Travis build instructions for new macOS VM image

jballanc committed:
Update Travis build instructions for new macOS VM image

jballanc committed:
Also install SWIG on macOS via Homebrew

jballanc committed:
Ensure that the HOMEBREW option is set for Travis macOS builds

jballanc committed:
Attempt to force OpenSSL installation on Travis

melissalinkert commented:
Closing as discussed in the context of upcoming work on #32.
